File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/DoNotLockWeakIdentityObjects.vb
Imports System
Imports System.Reflection
Imports System.Threading

Public Class DoNotLockWeakIdentityObjects

    Private synchronized As New Object()
    Private marshalByRefObject As MarshalByRefObject = New Timer(Nothing)
    Private marshalByRefObjectDerivate As New Timer(Nothing)
    Private executionEngineException As New ExecutionEngineException()
    Private outOfMemoryException As New OutOfMemoryException()
    Private stackOverflowException As New StackOverflowException()
    Private aString As String = "some value"
    Private memberInfo As MemberInfo = GetType(String).GetProperty("Length")
    Private parameterInfo As ParameterInfo = GetType(String).GetMethod("Equals").ReturnParameter
    Private thread As New Thread(DirectCast(Nothing, ThreadStart))

    Public Sub Test()
        SyncLock synchronized ' Compliant
        End SyncLock

        SyncLock marshalByRefObject ' Noncompliant {{Replace this lock on 'MarshalByRefObject' with a lock against an object that cannot be accessed across application domain boundaries.}}
            '    ^^^^^^^^^^^^^^^^^^
        End SyncLock
        SyncLock marshalByRefObjectDerivate ' Noncompliant {{Replace this lock on 'Timer' with a lock against an object that cannot be accessed across application domain boundaries.}}
        End SyncLock
        SyncLock executionEngineException   ' Noncompliant
        End SyncLock
        SyncLock outOfMemoryException       ' Noncompliant
        End SyncLock
        SyncLock stackOverflowException     ' Noncompliant
        End SyncLock
        SyncLock aString        ' Noncompliant
        End SyncLock
        SyncLock memberInfo     ' Noncompliant
        End SyncLock
        SyncLock parameterInfo  ' Noncompliant
        End SyncLock
        SyncLock thread         ' Noncompliant
        End SyncLock
    End Sub

End Class

File: analyzers/tests/SonarAnalyzer.Rules.Tests/Cases/DoNotLockWeakIdentityObjects.cs
using System;
using System.Reflection;
using System.Threading;

namespace Tests.Diagnostics
{
    class DoNotLockWeakIdentityObjects
    {
        private object synchronized = new object();
        private MarshalByRefObject marshalByRefObject = new Timer(null);
        private Timer marshalByRefObjectDerivate = new Timer(null);
        private ExecutionEngineException executionEngineException = new ExecutionEngineException();
        private OutOfMemoryException outOfMemoryException = new OutOfMemoryException();
        private StackOverflowException stackOverflowException = new StackOverflowException();
        private string aString = "some value";
        private MemberInfo memberInfo = typeof(string).GetProperty("Length");
        private ParameterInfo parameterInfo = typeof(string).GetMethod("Equals").ReturnParameter;
        private Thread thread = new Thread((ThreadStart)null);

        private void Test()
        {
            lock (synchronized) { } // Compliant

            lock (marshalByRefObject) { } // Noncompliant {{Replace this lock on 'MarshalByRefObject' with a lock against an object that cannot be accessed across application domain boundaries.}}
//                ^^^^^^^^^^^^^^^^^^
            lock (marshalByRefObjectDerivate) { } // Noncompliant {{Replace this lock on 'Timer' with a lock against an object that cannot be accessed across application domain boundaries.}}
            lock (executionEngineException) { } // Noncompliant
            lock (outOfMemoryException) { } // Noncompliant
            lock (stackOverflowException) { } // Noncompliant
            lock (aString) { } // Noncompliant
            lock (memberInfo) { } // Noncompliant
            lock (parameterInfo) { } // Noncompliant
            lock (thread) { } // Noncompliant
        }
    }
}
